services/src/main/java/org/keycloak/services/resources/admin/UserResource.java

```diff
@@ -169,11 +167,13 @@ public Response updateUser(final UserRepresentation rep) {
                 }
             }
 
-            Response response = validateUserProfile(user, rep, session);
+            UserProfile profile = session.getProvider(UserProfileProvider.class).forUpdate(USER_RESOURCE, rep.toAttributes(), user);
+
+            Response response = validateUserProfile(profile);
             if (response != null) {
                 return response;
             }
-            updateUserFromRep(user, rep, session, true);
+            updateUserFromRep(profile, user, rep, session, true);
```

Comment:
Do we really need separate validate and update methods here both calling UserProfile? I believe profile.update() performs validation first and throws validation exception, so one call should be enough

Reply:
As mentioned before. I'm trying to not change anything else but code specific to integrate with the user profile. I would say that refactoring other parts in the code base would be a second step after the SPI is integrated, and working.